Various minor widget performance optimizations

reaperrr opened the pull request:
Some presumably tiny, but also low-hanging fruits I decided to pick along with #18530.

reaperrr committed:
Cache SupportPowersWidget offsets at initialization

As well as overlay font.

reaperrr committed:
Cache indicator sprite in ResourceBarWidget at initialization

reaperrr committed:
Cache hue picker sprite in HueSliderWidget

abcdefg30 commented on OpenRA.Mods.Common/Widgets/ProgressBarWidget.cs:
```diff
@@ -46,14 +48,21 @@ protected ProgressBarWidget(ProgressBarWidget other)
+		{
+			base.Initialize(args);
+
+			var rb = RenderBounds;
```
This var only seems to be used once, so is this really necessary?

reaperrr commented:
Updated.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/SliderWidget.cs:
```diff
@@ -103,7 +117,7 @@ Rectangle ThumbRect
 				var rb = RenderBounds;
 				var width = rb.Height;
-				var height = rb.Height;
+				var height = width;
```
This looks really weird. Are we sure this is intentional? (Perhaps get a comment.)

reaperrr replied:
I see what you mean.
I think making the thumb square is intentional, but the width = height & height = width stuff looks weird indeed, I just didn't notice while making the change.

abcdefg30 reviewed (commented):
Code changes lgtm, but untested.

dragunoff commented:
I didn't spot any regressions.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/SliderWidget.cs:
```diff
@@ -102,8 +116,8 @@ Rectangle ThumbRect
 				var rb = RenderBounds;
-				var width = rb.Height;
 				var height = rb.Height;
+				var width = height; // Thumb is supposed to be square
```
Style nit: Comments should go on their own line.

reaperrr commented:
Updated.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/SliderWidget.cs:
```diff
@@ -116,27 +132,21 @@ public override void Draw()
 				var tickPos = new float2(
-					trackOrigin + (i * (trackRect.Width - (int)tick.Size.X) / (Ticks - 1)) - tick.Size.X / 2,
-					trackRect.Bottom);
+					trackOrigin + (i * (trackRectangle.Width - (int)tick.Size.X) / (Ticks - 1)) - tick.Size.X / 2,
```
This crashes when opening an actor edit panel in the map editor because `tick` is null for some reason:
```
OpenRA engine version {DEV_VERSION}
Tiberian Dawn mod version {DEV_VERSION}
on map b5bfd7d220923d64cd41892acf6b94c7fd56f10c (Deterring Democracy + by Psydev).
Date: 2020-08-23 11:05:58Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.NullReferenceException`: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   at OpenRA.Mods.Common.Widgets.SliderWidget.Draw() in \OpenRA\OpenRA.Mods.Common\Widgets\SliderWidget.cs:Line 138.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 455.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Widgets.Widget.DrawOuter() in \OpenRA\OpenRA.Game\Widgets\Widget.cs:Line 457.
   at OpenRA.Game.RenderTick() in \OpenRA\OpenRA.Game\Game.cs:Line 674.
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:Line 806.
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:Line 825.
   at OpenRA.Game.InitializeAndRun(String[] args) in \OpenRA\OpenRA.Game\Game.cs:Line 261.
   at OpenRA.Program.Main(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 33.
```

reaperrr commented:
> This crashes when opening an actor edit panel in the map editor because tick is null for some reason:

That's kind of puzzling. Other parts of that widget didn't work either (though they didn't crash).
I suspect the RenderBounds can change if the position depends on the actor, so I dropped that commit before I spend too much time debugging something that isn't worth the hassle.
I also dropped the commit that cached the min/max bar length in the ProgressBar widget, it's used in various places and in light of the above I didn't want to take that risk for something so minor.

The offsets in the SupportPowerWidget should be safe though, and both the ingame ResourceBar indicators in cnc and hue marks in the color picker show up correctly, so for now it seems those are safe, too.

But if any more issues pop up, I guess it would be safer to just drop this PR.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/84dad7ec3e574769d99ef19538d6e68bcb526108...da0a0be03cf9035588a50d1fb4162b6f8a54c328)
